python-packages/order_utils/src/zero_ex/abi_utils.py

```diff
@@ -0,0 +1,83 @@
+"""Ethereum ABI utilities.
+
+Builds on the eth-abi package, adding some convenience methods like those found
+in npmjs.com/package/ethereumjs-abi.  Ideally, all of this code should be
+pushed upstream into eth-abi.
+"""
+
+import re
+from typing import Any, List
+
+from mypy_extensions import TypedDict
+
+from eth_abi import encode_abi  # type: ignore
+from web3 import Web3  # type: ignore
+
+
+class MethodSignature(TypedDict, total=False):
+    """Object interface to an ABI method signature."""
+
+    method: str
+    args: List[str]
+
+
+def parse_signature(signature: str) -> MethodSignature:
+    """Parse a method signature into its constituent parts.
+
+    >>> parse_signature("ERC20Token(address)")
+    {'method': 'ERC20Token', 'args': ['address']}
+    """
+    matches = re.match(r"^(\w+)\((.+)\)$", signature)
+    if matches is None:
+        raise Exception("Invalid method signature")
+    return {"method": matches[1], "args": matches[2].split(",")}
+
+
+def elementary_name(name: str) -> str:
+    """Convert from short to canonical names; barely implemented.
+
+    Modeled after ethereumjs-abi's ABI.elementaryName(), but only implemented
+    to support our particular use case.
+
+    >>> elementary_name("address")
+    'address'
+    """
+    if name != "address":
+        raise Exception("Unsupported name " + name)
+    return name
+
+
+def event_id(name: str, types: List[str]) -> str:
+    """Return the Keccak-256 hash of the given method.
+
+    >>> event_id("ERC20Token", ["address"])
+    '0xf47261b06eedbfce68afd46d0f3c27c60b03faad319eaf33103611cf8f6456ad'
+    """
+    signature = name + "(" + ",".join(list(map(elementary_name, types))) + ")"
+    return Web3.sha3(text=signature).hex()
+
+
+def method_id(name: str, types: List[str]) -> str:
+    """Return the 4-byte method identifier.
+
+    >>> method_id("ERC20Token", ["address"])
+    '0xf47261b0'
+    """
+    return event_id(name, types)[0:10]
```

Comment:
Is `event_id` guaranteed to always be longer then 10 characters?

Reply:
Yes, because it's a Keccak-256 checksum, as returned from `Web3.sha3()`, which is always 64 hex digits, as suggested by some research on the algorithm, and as verified by anecdotally checking digest lengths for a few different message inputs (an empty string, and a 64-character string).